src/spiProcessor/lombok/spi/SpiProcessorPersistence.java

```diff
@@ -0,0 +1,148 @@
+	}
+	
+	static CharSequence readFilerResource(FileObject resource, Messager logger, String pathName) {
+		try {
+			// Eclipse can't handle getCharContent, so we must use a reader...
+			return tryWithReader(resource);
+		} catch (FileNotFoundException e) {
+			return null;
+		} catch (IOException e) {
+			if (
+				e.getClass().getName().equals("org.eclipse.core.internal.resources.ResourceException") &&
+				e.getMessage() != null &&
+				e.getMessage().endsWith("does not exist.")) {
+				
+				return null;
+			}
+			
+			logger.printMessage(Kind.ERROR, SpiProcessor.toErrorMsg(e, pathName));
+			return null;
+		} catch (Exception other) {
+			// otherwise, probably javac: Some versions don't support the `openReader` method.
+			try {
+				return resource.getCharContent(true);
+			} catch (FileNotFoundException e) {
+				return null;
+			} catch (IOException e) {
+				logger.printMessage(Kind.ERROR, SpiProcessor.toErrorMsg(e, pathName));
+				return null;
+			}
+		}
+	}
+	
+	private static CharSequence tryWithReader(FileObject resource) throws IOException {
+		StringBuilder sb = new StringBuilder();
+		Reader raw = resource.openReader(true);
+		try {
+			BufferedReader in = new BufferedReader(raw);
+			for (String line = in.readLine(); line != null; line = in.readLine()) sb.append(line).append('\n');
+			return sb;
+		} finally {
+			if (raw != null) raw.close();
+		}
+	}
+	
+	Collection<String> tryFind() {
+		File dir = determineOutputLocation();
+		if (dir == null || !dir.isDirectory()) return Collections.emptyList();
+		List<String> out = new ArrayList<String>();
+		for (File p : dir.listFiles()) {
+			if (!p.isFile()) continue;
+			out.add(p.getName());
+		}
+		return out;
+	}
+	
+	private File determineOutputLocation() {
+		FileObject resource;
+		try {
+			resource = filer.createResource(StandardLocation.CLASS_OUTPUT, "META-INF", "locator.tmp");
+		} catch (FileNotFoundException e) {
+			e.printStackTrace();
+			// Could happen
+			return null;
+		} catch (IOException e) {
+			e.printStackTrace();
+			logger.printMessage(Kind.NOTE, "IOException while determining output location: " + e.getMessage());
+			return null;
+		} catch (IllegalArgumentException e) {
+			e.printStackTrace();
+			// Happens when the path is invalid. For instance absolute or relative to a path 
+			// not part of the class output folder.
+			//
+			// Due to a bug in javac for Linux, this also occurs when no output path is specified 
+			// for javac using the -d parameter.
+			// See http://forums.sun.com/thread.jspa?threadID=5240999&tstart=45
+			// and http://bugs.sun.com/bugdatabase/view_bug.do?bug_id=6647996
+			
+			return null;
+		}
+		
+		URI uri = resource.toUri();
+		return new File(new File(uri).getParentFile(), "services");
+	}
+	
+	void write(String serviceName, String value) throws IOException {
+		FileObject output = filer.createResource(StandardLocation.CLASS_OUTPUT, "", path + serviceName);
+		Writer writer = output.openWriter();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *writer*. The resource is referenced at lines 143, 144, 145. The resource is closed at line 146. There are other execution paths that do not contain closure statements, for example, when Writer.write throws an exception. Ensure *writer* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).